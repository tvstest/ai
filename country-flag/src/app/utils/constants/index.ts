export const COUNTRIES_API_BASE_URL = process.env.REACT_APP_COUNTRY_API;
export const WEATHER_API_BASE_URL = process.env.REACT_APP_WEATHER_API;
export const WEATHER_API_ACCESS_KEY = process.env.REACT_APP_WEATHER_API_KEY;
export const ERROR_FETCHING_WEATHER =
    "Some error occurred while fetching weather!";
export const ERROR_FETCHING_COUNTRIES = "No countries found!";