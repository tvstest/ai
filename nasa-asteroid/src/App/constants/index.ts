export const NASA_BASE_URL = 'https://api.nasa.gov/neo/rest/v1/neo/'
export const NASA_API_KEY_PREFIX = `?api_key=${process.env.REACT_APP_NASA_API_KEY}`
