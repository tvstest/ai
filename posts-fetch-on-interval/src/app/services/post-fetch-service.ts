import { AxiosResponse } from 'axios'
import { BASE_URL } from 'app/configs'
import { IPostFetchData } from 'app/utility/interface/post-data'
import httpClient from './base-service'

export const GetPostsData = (
  page: number
): Promise<AxiosResponse<IPostFetchData>> =>
  httpClient.get(`${BASE_URL}?tags=story&page=${page}`)
