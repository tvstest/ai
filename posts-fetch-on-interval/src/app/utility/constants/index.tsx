export const INTERNAL_SERVER_ERROR = 'Internal Server Error'
export const SOMETHING_WENT_WRONG = 'Something Went Wrong.'
export const ALL_DATA_FETCHED = 'All posts have been fetched.'
export const COULD_NOT_LOAD_DATA = 'Could not load list.'
export const TIMER_TEST_SECONDS = 21000
export const TIMER_MOCK_TIMEOUT = 40000
