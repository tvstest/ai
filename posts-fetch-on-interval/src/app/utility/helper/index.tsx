export const showLoader = (): void => {
  const loaderDiv = document.getElementById('loaderForAPICall')
  if (loaderDiv) {
    loaderDiv.classList.add('loaderShow')
    loaderDiv.classList.remove('loaderHide')
  }
}

export const hideLoader = (): void => {
  const loaderDiv = document.getElementById('loaderForAPICall')
  if (loaderDiv) {
    loaderDiv.classList.remove('loaderShow')
    loaderDiv.classList.add('loaderHide')
  }
}

export const sleep = (m: number): Promise<NodeJS.Timeout> =>
  new Promise((r) => setTimeout(r, m))
