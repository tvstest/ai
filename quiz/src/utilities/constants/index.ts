import { Language } from 'utilities/enum/language'

export const DEFAULT_QUESTION_INDEX = 0

export const GENDER = [
  { label: 'Male', value: 'male' },
  { label: 'Female', value: 'female' },
]

export const LANGUAGE = [
  { label: Language[Language.English], value: Language.English },
  { label: Language[Language.Hindi], value: Language.Hindi },
]

export const COLORS = ['#2e7d32', '#d32f2f', '#ed6c02']
export const RADIAN = Math.PI / 180
