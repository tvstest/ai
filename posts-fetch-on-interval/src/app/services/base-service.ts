import axios, { AxiosError, AxiosRequestConfig, AxiosResponse } from 'axios'
import { toast } from 'react-toastify'
import { HttpStatusCodes } from 'app/utility/enums/http-status-codes'
import { hideLoader, showLoader } from 'app/utility/helper'
import {
  SOMETHING_WENT_WRONG,
  INTERNAL_SERVER_ERROR,
} from 'app/utility/constants'

declare module 'axios' {
  export interface AxiosRequestConfig {
    hideLoader?: boolean
  }
}

axios.interceptors.request.use(
  async (config: AxiosRequestConfig) => {
    if (!config.hideLoader) {
      showLoader()
    }
    return config
  },
  (error: AxiosError) => {
    return Promise.reject(error)
  }
)

axios.interceptors.response.use(
  (response: AxiosResponse) => {
    hideLoader()
    return response
  },
  (error: AxiosError) => {
    const expectedError =
      error.response &&
      error.response.status >= HttpStatusCodes.BadRequest &&
      error.response.status < HttpStatusCodes.InternalServerError

    if (expectedError) {
      if (!error.response?.data?.success && error.response?.data?.message) {
        toast.error(INTERNAL_SERVER_ERROR)
      }
    } else {
      toast.error(SOMETHING_WENT_WRONG)
    }
    hideLoader()

    return Promise.reject(error)
  }
)

export default {
  post: axios.post,
  get: axios.get,
}
