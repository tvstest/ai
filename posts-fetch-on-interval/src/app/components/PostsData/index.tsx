import React, { useEffect, useState, useRef, useCallback } from 'react'
import { useInView } from 'react-intersection-observer'
import { toast } from 'react-toastify'
import {
  TableContainer,
  Table,
  TableBody,
  TableHead,
  TableCell,
  TableRow,
  Box,
  Typography,
  TextField,
} from '@mui/material'
import { makeStyles } from '@mui/styles'
import Autocomplete from '@mui/material/Autocomplete'
import { IPostFetchHitsData } from 'app/utility/interface/post-data'
import { GetPostsData } from 'app/services/post-fetch-service'
import { ALL_DATA_FETCHED, COULD_NOT_LOAD_DATA } from 'app/utility/constants'
import { FETCH_DATA_INTERVAL_TIME } from 'app/configs'
import DialogBox from '../DialogBox'
import DatePicker from '../common/DatePickerComponent'

const useStyles = makeStyles({
  root: {
    flexGrow: 1,
    width: '100%',
    height: '100%',
    overflow: 'auto',
    paddingBottom: '2%',
  },
  tableCell: {
    border: '1px solid #aaa',
    maxWidth: '400px',
    overflowWrap: 'break-word',
  },
  tableRowStriped: {
    background: '#3f51b596',
    cursor: 'pointer',
  },
  tableRow: {
    cursor: 'pointer',
    background: '#ffffff',
  },
  tableHeadCell: {
    textAlign: 'center',
    border: '1px solid #aaa',
    position: 'sticky',
    fontWeight: 600,
  },
  filterDiv: {
    margin: '0px !important',
  },
  searchContainer: {
    paddingBottom: '2%',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'space-between',
    margin: '3% 0',
  },
  postListContainer: {
    padding: '16px',
  },
  searchInput: {
    width: '100%',
    background: '#ffffff',
    borderRadius: '5px',
  },
  noData: {
    display: 'flex',
    width: '100%',
    justifyContent: 'center',
    alignItems: 'center',
  },
  autocomplete: {
    background: '#ffffff',
    borderRadius: '5px',
  },
})

interface ITableValueProps {
  value: string
}

const columns = [
  {
    id: 'title',
    label: 'Title',
  },
  {
    id: 'author',
    label: 'Author',
  },
  {
    id: 'url',
    label: 'URL',
    format: (value: ITableValueProps) => value || '-',
  },
  {
    id: 'created_at',
    label: 'Created Date',
    format: (value: ITableValueProps) => value || '-',
  },
]

const PostsData = React.memo(() => {
  const classes = useStyles()
  const [allPostData, setPostData] = useState<IPostFetchHitsData[]>([])
  const [allPostsFetched, setAllPostsFetched] = useState(false)
  const [selectedRowData, setSelectedRowData] = useState<IPostFetchHitsData>()
  const [open, setOpen] = useState(false)
  const [search, setSearch] = useState('')
  const [title, setTitle] = useState('')
  const [autoCompleteInput, setAutoCompleteInput] = useState('')
  const [selectedDate, setSelectedDate] = useState(null)
  const currentPage = useRef<number>(null)
  const [lastPost, isLastPostVisible] = useInView()

  const postEndingRef = useCallback(
    (node) => {
      lastPost(node)
    },
    [lastPost]
  )

  const handleGetPostList = useCallback(async () => {
    const page: number = currentPage.current ? currentPage.current : 0
    currentPage.current = page + 1
    try {
      const res = await GetPostsData(page)
      const dataList: IPostFetchHitsData[] = res.data.hits
      if (page >= res.data.nbPages) {
        setAllPostsFetched(true)
        toast.info(ALL_DATA_FETCHED)
      }
      setPostData((existingPosts) => [
        ...existingPosts.filter(
          (item: IPostFetchHitsData) =>
            !dataList.some(
              (i: IPostFetchHitsData) => i.objectID === item.objectID
            )
        ),
        ...dataList,
      ])
    } catch (err) {
      toast.error(COULD_NOT_LOAD_DATA)
    }
  }, [])

  const handleOpenDetailPopup = () => {
    setOpen(true)
  }

  const handleCloseDetailPopup = () => {
    setOpen(false)
  }

  const handleRowClick = useCallback((data: IPostFetchHitsData) => {
    setSelectedRowData(data)
    handleOpenDetailPopup()
  }, [])

  useEffect(() => {
    let unmounted = true
    if (
      unmounted &&
      isLastPostVisible &&
      allPostData?.length > 0 &&
      !allPostsFetched
    ) {
      handleGetPostList()
    }
    return () => {
      unmounted = false
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isLastPostVisible])

  useEffect(() => {
    let fetchData: any
    if (!allPostsFetched) {
      handleGetPostList()
      fetchData = setInterval(() => {
        handleGetPostList()
      }, FETCH_DATA_INTERVAL_TIME)
    }
    return () => {
      clearInterval(fetchData)
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [allPostsFetched])

  const handleSearch = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSearch(e.target.value)
  }

  const handleChangeDate = (e: React.ChangeEvent<HTMLInputElement>) => {
    setSelectedDate(e.target.value)
  }

  let filteredPosts: IPostFetchHitsData[] = [...allPostData]

  if (search.trim()) {
    filteredPosts = filteredPosts.filter(
      (item) =>
        item.author?.toLowerCase().includes(search.trim().toLowerCase()) ||
        item.url?.toLowerCase().includes(search.trim().toLowerCase()) ||
        item.title?.toLowerCase().includes(search.trim().toLowerCase())
    )
  }

  if (selectedDate) {
    filteredPosts = filteredPosts.filter((item) => {
      const itemCreatedDate = new Date(item.created_at)
        .toISOString()
        .slice(0, 10)

      return selectedDate === itemCreatedDate
    })
  }

  if (autoCompleteInput.trim()) {
    filteredPosts = filteredPosts.filter(
      (item) => item.title?.toLowerCase() === autoCompleteInput?.toLowerCase()
    )
  }

  return (
    <Box className={classes.root}>
      <Box className={classes.postListContainer}>
        <Box className={classes.searchContainer}>
          <TextField
            id="search"
            inputProps={{ 'data-testid': 'search' }}
            label="Search post"
            variant="outlined"
            name="search"
            value={search}
            onChange={handleSearch}
            className={classes.searchInput}
          />
        </Box>
        <Box className={classes.searchContainer}>
          <>
            <Autocomplete
              freeSolo
              disableClearable
              fullWidth
              value={title}
              options={allPostData.map((post) => post.title)}
              onChange={(event: any, newValue: string | null) => {
                setTitle(newValue)
              }}
              onInputChange={(event: any, newValue: string | null) => {
                setAutoCompleteInput(newValue)
              }}
              renderInput={(params) => (
                <TextField
                  {...params}
                  label="Title"
                  variant="outlined"
                  InputProps={{ ...params.InputProps, type: 'search' }}
                  className={classes.autocomplete}
                />
              )}
            />
            <DatePicker
              handleChangeDate={(event: React.ChangeEvent<HTMLInputElement>) =>
                handleChangeDate(event)
              }
            />
          </>
        </Box>

        {filteredPosts && filteredPosts.length > 0 ? (
          <Box>
            <TableContainer className={classes.filterDiv}>
              <Table aria-label="sticky table">
                <TableHead>
                  <TableRow>
                    {columns.map((column) => (
                      <TableCell
                        className={classes.tableHeadCell}
                        key={column.id}
                      >
                        {column.label}
                      </TableCell>
                    ))}
                  </TableRow>
                </TableHead>
                <TableBody>
                  {filteredPosts.map((item: any, rowIndex) => (
                    <TableRow
                      hover
                      className={
                        rowIndex % 2
                          ? classes.tableRowStriped
                          : classes.tableRow
                      }
                      tabIndex={-1}
                      key={`${item.objectID}`}
                      onClick={() => handleRowClick(item)}
                    >
                      {columns.map((column) => {
                        const value = item[column.id]
                        return (
                          <TableCell
                            className={classes.tableCell}
                            key={`${item.objectID}-${column.id}`}
                          >
                            {column.format ? column.format(value) : value}
                          </TableCell>
                        )
                      })}
                    </TableRow>
                  ))}
                </TableBody>
              </Table>
            </TableContainer>
          </Box>
        ) : (
          <Box>
            <div className={classes.noData}>
              <Typography variant="body1"> No data found </Typography>
            </div>
          </Box>
        )}
        <div ref={postEndingRef} style={{ opacity: '0' }}>
          End of data
        </div>
      </Box>
      <DialogBox
        open={open}
        close={handleCloseDetailPopup}
        data={selectedRowData}
      />
    </Box>
  )
})

export default PostsData
