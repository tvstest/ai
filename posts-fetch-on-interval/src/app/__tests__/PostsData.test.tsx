import React from 'react'
import { render, queryByAttribute, act, waitFor } from '@testing-library/react'
import renderer from 'react-test-renderer'
import { setupIntersectionObserverMock } from 'app/utility/testUtilities'
import * as PostFetchServices from 'app/services/post-fetch-service'
import PostsData from 'app/components/PostsData/index'
import { TIMER_TEST_SECONDS, TIMER_MOCK_TIMEOUT } from 'app/utility/constants'
import { sleep } from 'app/utility/helper'

beforeEach(() => {
  setupIntersectionObserverMock()
})

test('renders posts table data', () => {
  const getById = queryByAttribute.bind(null, 'id')
  const dom = render(<PostsData />)
  const table = getById(dom.container, 'detail-table')
  expect(table).toBeDefined()
})

test('input elements are rendered properly', async () => {
  const { getByTestId } = render(<PostsData />)
  const inputElement = getByTestId('search')
  expect(inputElement).toBeInTheDocument()
})

test('Get posts api should be called once only on component render', async () => {
  const mockApiFunction = jest.spyOn(PostFetchServices, 'GetPostsData')
  await act(async () => {
    render(<PostsData />)
    await waitFor(async () => {
      expect(mockApiFunction).toHaveBeenCalledTimes(1)
    })
  })
})

test(
  'After TIMER_TEST_SECONDS time interval Get posts api should be called thrice',
  async () => {
    const mockApiFunction = jest.spyOn(PostFetchServices, 'GetPostsData')
    await act(async () => {
      render(<PostsData />)
      await sleep(TIMER_TEST_SECONDS)
      await waitFor(async () => {
        expect(mockApiFunction).toHaveBeenCalledTimes(3)
      })
    })
  },
  TIMER_MOCK_TIMEOUT
)

test('renders PostData correctly', () => {
  const domTree = renderer.create(<PostsData />).toJSON()
  expect(domTree).toMatchSnapshot()
})
