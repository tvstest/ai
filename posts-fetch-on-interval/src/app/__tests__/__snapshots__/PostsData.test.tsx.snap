// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PostData correctly 1`] = `
<div
  className="makeStyles-root-69 MuiBox-root css-0"
>
  <div
    className="makeStyles-postListContainer-76 MuiBox-root css-0"
  >
    <div
      className="makeStyles-searchContainer-75 MuiBox-root css-0"
    >
      <div
        className="MuiFormControl-root MuiTextField-root makeStyles-searchInput-77 css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={false}
          htmlFor="search"
          id="search-label"
        >
          Search post
        </label>
        <div
          className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            data-testid="search"
            disabled={false}
            id="search"
            name="search"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="text"
            value=""
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-173wfxe"
            >
              <span>
                Search post
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      className="makeStyles-searchContainer-75 MuiBox-root css-0"
    >
      <div
        aria-expanded={false}
        aria-owns={null}
        className="MuiAutocomplete-root MuiAutocomplete-fullWidth css-eozd4h-MuiAutocomplete-root"
        onClick={[Function]}
        onKeyDown={[Function]}
        onMouseDown={[Function]}
        role="combobox"
      >
        <div
          className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root makeStyles-autocomplete-79 css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            id="undefined-label"
          >
            Title
          </label>
          <div
            className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd MuiAutocomplete-inputRoot css-154xyx0-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <input
              aria-activedescendant={null}
              aria-autocomplete="list"
              aria-controls={null}
              aria-invalid={false}
              autoCapitalize="none"
              autoComplete="off"
              autoFocus={false}
              className="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputTypeSearch MuiInputBase-inputAdornedEnd MuiAutocomplete-input MuiAutocomplete-inputFocused css-1j5luls-MuiInputBase-input-MuiOutlinedInput-input"
              disabled={false}
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onMouseDown={[Function]}
              required={false}
              spellCheck="false"
              type="search"
              value=""
            />
            <div
              className="MuiAutocomplete-endAdornment css-1q60rmi-MuiAutocomplete-endAdornment"
            />
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-173wfxe"
              >
                <span>
                  Title
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root makeStyles-textField-80 css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          className="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-1kty9di-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink={true}
          htmlFor="date"
          id="date-label"
        >
          Created At
        </label>
        <div
          className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            defaultValue=""
            disabled={false}
            id="date"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={false}
            type="date"
          />
          <fieldset
            aria-hidden={true}
            className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              className="css-1z7n62"
            >
              <span>
                Created At
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      className="MuiBox-root css-0"
    >
      <div
        className="makeStyles-noData-78"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
        >
           No data found 
        </p>
      </div>
    </div>
    <div
      style={
        Object {
          "opacity": "0",
        }
      }
    >
      End of data
    </div>
  </div>
</div>
`;
